probe/awsecs/client.go

```diff
@@ -129,85 +129,67 @@ func (c ecsClientImpl) getCachedService(serviceName string) (EcsService, bool) {
 	return EcsService{}, false
 }
 
-// Returns a channel from which service ARNs can be read.
+// Returns a list of service names.
 // Cannot fail as it will attempt to deliver partial results, though that may end up being no results.
-func (c ecsClientImpl) listServices() <-chan string {
+func (c ecsClientImpl) listServices() []string {
 	log.Debugf("Listing ECS services")
-	results := make(chan string)
-	go func() {
-		count := 0
-		err := c.client.ListServicesPages(
-			&ecs.ListServicesInput{Cluster: &c.cluster},
-			func(page *ecs.ListServicesOutput, lastPage bool) bool {
-				for _, arn := range page.ServiceArns {
-					count++
-					results <- *arn
-				}
-				return true
-			},
-		)
-		if err != nil {
-			log.Warnf("Error listing ECS services, ECS service report may be incomplete: %v", err)
-		}
-		log.Debugf("Listed %d services", count)
-		close(results)
-	}()
+	results := []string{}
+	err := c.client.ListServicesPages(
+		&ecs.ListServicesInput{Cluster: &c.cluster},
+		func(page *ecs.ListServicesOutput, lastPage bool) bool {
+			for _, name := range page.ServiceArns {
+				results = append(results, *name)
+			}
+			return true
+		},
+	)
+	if err != nil {
+		log.Warnf("Error listing ECS services, ECS service report may be incomplete: %v", err)
+	}
+	log.Debugf("Listed %d services", len(results))
 	return results
 }
 
-// Returns (input, done) channels. Service ARNs given to input are batched and details are fetched,
-// with full EcsService objects being put into the cache. Closes done when finished.
-func (c ecsClientImpl) describeServices() (chan<- string, <-chan struct{}) {
-	input := make(chan string)
-	done := make(chan struct{})
+// Service names given are batched and details are fetched,
+// with full EcsService objects being put into the cache.
+// Cannot fail as it will attempt to deliver partial results.
+func (c ecsClientImpl) describeServices(services []string) {
+	const maxServices = 10 // How many services we can put in one Describe command
+	group := sync.WaitGroup{}
 
 	log.Debugf("Describing ECS services")
 
-	go func() {
-		const maxServices = 10 // How many services we can put in one Describe command
-		group := sync.WaitGroup{}
-
-		// count and calls is just for logging
-		count := 0
-		calls := 0
-
-		batch := make([]string, 0, maxServices)
-		for arn := range input {
-			batch = append(batch, arn)
-			if len(batch) == maxServices {
-				group.Add(1)
-				go func(arns []string) {
-					defer group.Done()
-					c.describeServicesBatch(arns)
-				}(batch)
-				count += len(batch)
-				calls++
-				batch = make([]string, 0, maxServices)
-			}
-		}
-		if len(batch) > 0 {
-			c.describeServicesBatch(batch)
-			count += len(batch)
-			calls++
-		}
+	// split into batches
+	batches := make([][]string, 0, len(services)/maxServices+1)
```

Comment:
I think it would be event simpler to spawn the goroutines directly (factoring the spawning into a function called inside the loop) instead of accumulating the batches in slices of slices.